File: 155-MinStack/155-MinStack.java
// Last updated: 8/13/2025, 8:21:52 PM
class MinStack {

    Stack<Long> s = new Stack();
    long min;

    public MinStack() {
        
    }
    
    public void push(int val) {
        if(s.size()==0){
            s.add((long)0);
            min = (long)val;
        }
        else{
            long diff = val - min;
            s.add(diff);
            if(diff<0){
                min = (long)val;
            }
        }
    }
    
    public void pop() {
        long val;
        if(s.peek()<0){
            val = s.pop();
            min = (long)min - val;
        }
        else{
            s.pop();
        }
    }
    
    public long top() {
        if(s.peek()<0){
            return (int)min;
        }
        return s.peek() + min;
    }
    
    public int getMin() {
        return (int)min;
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(val);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */

File: 155-MinStack/155-MinStack.py
# Last updated: 8/13/2025, 8:21:46 PM
class MinStack:

    def __init__(self):
        self.stack = []
        self.current_min = float('inf')

    def push(self, val: int) -> None:
        if len(self.stack) == 0 : 
            self.current_min = val
        diff = val - self.current_min
        self.stack.append(diff)
        if diff < 0 : 
            self.current_min = val

    def pop(self) -> None:
        popped = self.stack.pop(-1)
        if popped < 0 : 
            value = self.current_min
            self.current_min = self.current_min - popped
            return value
        return self.current_min + popped

    def top(self) -> int:
        top_value = self.stack[-1]
        if top_value < 0 : 
            return self.current_min
        else : 
            return self.current_min + top_value

    def getMin(self) -> int:
        return self.current_min


# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(val)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()